pkg/autodiscovery/providers/prometheus_services.go

```diff
@@ -20,42 +21,104 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/labels"
 	listersv1 "k8s.io/client-go/listers/core/v1"
 	"k8s.io/client-go/tools/cache"
 )
 
+type ServiceAPI interface {
+	// List lists all Services
+	ListServices() ([]*v1.Service, error)
+	// GetEndpoints gets Endpoints by namespace and name
+	GetEndpoints(namespace, name string) (*v1.Endpoints, error)
+}
+
+type svcAPI struct {
+	serviceLister   listersv1.ServiceLister
+	endpointsLister listersv1.EndpointsLister
+}
+
+func (api *svcAPI) ListServices() ([]*v1.Service, error) {
+	return api.serviceLister.List(labels.Everything())
+}
+
+func (api *svcAPI) GetEndpoints(namespace, name string) (*v1.Endpoints, error) {
+	return api.endpointsLister.Endpoints(namespace).Get(name)
+}
+
 // PrometheusServicesConfigProvider implements the ConfigProvider interface for prometheus services
 type PrometheusServicesConfigProvider struct {
-	lister   listersv1.ServiceLister
+	sync.RWMutex
+
+	api      ServiceAPI
 	upToDate bool
-	PrometheusConfigProvider
+
+	collectEndpoints   bool
+	monitoredEndpoints map[string]bool
+
+	checks []*common.PrometheusCheck
 }
 
 // NewPrometheusServicesConfigProvider returns a new Prometheus ConfigProvider connected to kube apiserver
-func NewPrometheusServicesConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
+func NewPrometheusServicesConfigProvider(configProviders config.ConfigurationProviders) (ConfigProvider, error) {
 	ac, err := apiserver.GetAPIClient()
 	if err != nil {
 		return nil, fmt.Errorf("cannot connect to apiserver: %s", err)
 	}
 
 	servicesInformer := ac.InformerFactory.Core().V1().Services()
 	if servicesInformer == nil {
-		return nil, errors.New("cannot get service informer")
+		return nil, errors.New("cannot get services informer")
+	}
+
+	var endpointsLister listersv1.EndpointsLister
+
+	collectEndpoints := config.Datadog.GetBool("prometheus_scrape.service_endpoints")
+	endpointsInformer := ac.InformerFactory.Core().V1().Endpoints()
+	if collectEndpoints {
+		if endpointsInformer == nil {
+			return nil, errors.New("cannot get endpoints informer")
+		}
+		endpointsLister = endpointsInformer.Lister()
+
+	}
+
+	api := &svcAPI{
+		serviceLister:   servicesInformer.Lister(),
+		endpointsLister: endpointsLister,
 	}
 
-	p := &PrometheusServicesConfigProvider{
-		lister: servicesInformer.Lister(),
+	// TODO: refactor PrometheusConfigProvider to be an init once helper
+	configProvider := &PrometheusConfigProvider{}
+	err = configProvider.setupConfigs()
+	if err != nil {
+		return nil, err
 	}
 
+	p := newPromServicesProvider(configProvider.checks, api, collectEndpoints)
+
 	servicesInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    p.invalidate,
 		UpdateFunc: p.invalidateIfChanged,
 		DeleteFunc: p.invalidate,
 	})
 
-	err = p.setupConfigs()
-	return p, err
+	if collectEndpoints {
```

Comment:
can we have a single `if collectEndpoints` block in this function?

Reply:
This is a bit tricky here due to the initialization sequence - 1) we initialize dependencies for the provider first (checking the boolean) and then 2) we initialize the provider 2) we set up invalidate hooks (conditional on the boolean).